axum/src/extract/host.rs

```diff
@@ -0,0 +1,59 @@
+use super::{FromRequest, RequestParts};
+use async_trait::async_trait;
+use std::{convert::Infallible};
+
+/// Extractor that extracts the host from a request.
+#[derive(Debug, Clone, Default)]
+pub struct Host(pub String);
+
+#[async_trait]
+impl<B> FromRequest<B> for Host
+where
+    B: Send,
+{
+    type Rejection = Infallible;
+
+    async fn from_request(req: &mut RequestParts<B>) -> Result<Self, Self::Rejection> {
+        if let Some(host) = req.uri().host() {
+            return Ok(Host(host.to_string()));
+        }
+
+        if let Some(Ok(host)) = req.headers().get("host").map(|host| host.to_str()) {
+            return Ok(Host(host.to_string()));
+        }
```

Comment:
Should this extractor also handles the `x-forwarded-host` header?

Reply:
Yes I think so.

Prior art: actix-web looks at that https://docs.rs/actix-web/latest/actix_web/dev/struct.ConnectionInfo.html#method.host, as well as `Forwarded`.

That might involve a bunch of header parsing which can be hard so I'd be fine with adding this in a follow up PR.

Reply:
In the actix documentation they say they look at uri after all the headers, however I've done the opposite. If we're adopting the actix header suggestions should we also use the same ordering?

Reply:
I wonder is there some canonical/standard for resolving the host of a http request?

Reply:
It would be cool if the [headers](https://docs.rs/headers/latest/headers) crate supported it but that doesn't seem to be the case.